mixins/selectable/selectable-mixin.js

```diff
@@ -0,0 +1,89 @@
+import { css, html } from 'lit-element/lit-element.js';
+import { checkboxStyles } from '../../components/inputs/input-checkbox-styles.js';
+import { getUniqueId } from '../../helpers/uniqueId.js';
+import { nothing } from 'lit-html';
+
+export const SelectableMixin = superclass => class extends superclass {
+
+	static get properties() {
+		return {
+			disabled: { type: Boolean },
+			key: { type: String, reflect: true },
+			selectable: {type: Boolean },
+			selected: { type: Boolean, reflect: true }
+		};
+	}
+
+	static get styles() {
+		return [ checkboxStyles, css`
+			.d2l-select-action {
+				height: 100%;
+				display: block;
+				cursor: pointer;
+			}
+			.d2l-select-action[disabled] {
+				cursor: default;
+			}
+		` ];
+	}
+
+	constructor() {
+		super();
+		this._checkboxId = getUniqueId();
+	}
+
+	connectedCallback() {
+		super.connectedCallback();
+		this._selectedEventName = `d2l-${this.constructor.name
+			.replace(/([a-z0-9])([A-Z])/g, '$1-$2').toLowerCase()}-selected`;
+		if (!this.key) {
+			console.warn('"key" property has not been set on selectable item');
+			this.setSelected(undefined, true);
+		}
+	}
+
+	setSelected(selected, suppressEvent = false) {
+		this.selected = selected;
+		if (!suppressEvent) this._dispatchSelected(selected);
+	}
+
+	_dispatchSelected(value) {
+		this.dispatchEvent(new CustomEvent(this._selectedEventName, {
+			detail: { key: this.key, selected: value },
+			bubbles: true
+		}));
+	}
+
+	_handleCheckboxChange(event) {
+		this.setSelected(event.target.checked);
+	}
+
+	_handleSelectActionClick(event) {
+		event.preventDefault();
+		if (this.disabled) return;
+		this.setSelected(!this.selected);
+	}
+
+	_renderCheckbox() {
+		return this.selectable ? html`
+			<input
+				id="${this._checkboxId}"
+				class="d2l-input-checkbox"
+				@change="${this._handleCheckboxChange}"
+				type="checkbox"
+				.checked="${this.selected}"
+				?disabled="${this.disabled}">
+			` : nothing;
+	}
+
+	_renderSelectAction(inner) {
+		return this.selectable ? html`
+			<label @click="${this._handleSelectActionClick}"
+				class="d2l-select-action"
+				?disabled="${this.disabled}"
+				for="${this._checkboxId}">
+				${inner}
```

Comment:
what is purpose of the inner ?

Reply:
It is an optional argument that is used to place text or html or other components within the `label`, but it should probably be using `nothing` instead to prevent an empty text node

Reply:
nothing is for slots to default. so null is good

Reply:
Be a little careful here. `<label>` elements should only be used to wrap or link to inputs. And even then, we ran into a lot of issues with date/time where we had a label wrapping a text input and AT ends up reading a lot of things you don't want it to. Is it expected that they'll pass `_renderCheckbox` in as `inner`? Because then things get double-labelled, which is weird.

Also... if they don't pass anything for `inner`, then the label that's linked to the checkbox won't have any text associated with it, so the checkbox will be unlabelled, which we can't have.

Reply:
I've been contemplating adding an `aria-labelledby` to the `label` for cases where `inner` isn't passed, which I'm realizing now is a necessary thing.

If users decide to pass `_renderCheckbox` as `inner`, is that bad because the `label` already has the `for` attribute associated with the checkbox?